File: MedianOfTwoSortedArrays/MedianOfTwoSortedArrays.cpp
/*
 * explanation: http://blog.unieagle.net/2012/10/04/leetcode%E9%A2%98%E7%9B%AE%EF%BC%9Amedian-of-two-sorted-arrays/
 * time: O(logk), space: O(1)
 */
#include <iostream>
#include <cassert>
using namespace std;

class Solution{
public:
    int findKth(int A[], int m, int B[], int n, int k){
        if(m == 0)
            return B[k - 1];
        if(n == 0)
            return A[k - 1];
        if(k == 1)
            return min(A[0], B[0]);
        
        int ka, kb;
        if(m < n){
            ka = min(k / 2, m);
            kb = k - ka;
        }
        else{
            kb = min(k / 2, n);
            ka = k - kb;
        }
        
        if(A[ka - 1] < B[kb - 1])
            return findKth(A + ka, m - ka, B, n, k - ka);
        else
            return findKth(A, m, B + kb, n - kb, k - kb);
    }
    
    double findMedianSortedArrays(int A[], int m, int B[], int n){
        int t = m + n;
        assert(t > 0);
        
        if(t % 2)
            return findKth(A, m, B, n, (t + 1) / 2);
        else
            return (findKth(A, m, B, n, t / 2) + findKth(A, m, B, n, t / 2 + 1)) / 2.0;
    }
};

int main(){
    int A[] = {1, 3, 5, 7};
    int B[] = {2, 4, 6, 8};
    int m = sizeof(A) / sizeof(A[0]);
    int n = sizeof(B) / sizeof(B[0]);
    
    Solution solution;
    cout << solution.findMedianSortedArrays(A, m, B, n) << endl;
    
    return 0;
}

File: algorithms/MedianOfTwoSortedArrays/MedianOfTwoSortedArrays.py
class Solution(object):
    def findKth(self, nums1, nums2, k):
        n1, n2 = len(nums1), len(nums2)
        if n1 == 0:
            return nums2[k - 1]
        if n2 == 0:
            return nums1[k - 1]
        if k == 1:
            return min(nums1[0], nums2[0])
        
        k1, k2 = None, None
        if n1 < n2:
            k1 = min(k / 2, n1)
            k2 = k - k1
        else:
            k2 = min(k / 2, n2)
            k1 = k - k2
            
        if nums1[k1 - 1] < nums2[k2 - 1]:
            return self.findKth(nums1[k1:], nums2, k - k1)
        else:
            return self.findKth(nums1, nums2[k2:], k - k2)
        
    def findMedianSortedArrays(self, nums1, nums2):
        n1, n2 = len(nums1), len(nums2)
        n = n1 + n2
        
        if(n % 2):
            return self.findKth(nums1, nums2, (n + 1) / 2)
        else:
            return (self.findKth(nums1, nums2, n / 2) + self.findKth(nums1, nums2, n / 2 + 1)) / 2.0
